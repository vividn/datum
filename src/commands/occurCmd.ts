import { handleTimeArgs, timeArgs, TimeArgs } from "../input/timeArgs";
import { addCmd, AddCmdArgs, newDocArgs } from "./addCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { dataArgs } from "../input/dataArgs";
import { fieldArgs } from "../input/fieldArgs";

export const occurArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, newDocArgs, timeArgs, dataArgs],
});
export const occurCmdArgs = new ArgumentParser({
  description: "record the occurence of something at a single time point",
  prog: "datum occur",
  usage: `%(prog)s <field> [data..]
  %(prog)s --fieldless [data..]
  %(prog)s <field> -K <reqKey1> ... -K <reqKeyN> -k <optKey1>[=defaultVal1] ... -k <optKeyN> <reqVal1> ... <reqValN> [optVal1] ... [optValN] [data..]
`,
  parents: [occurArgs, dbArgs, outputArgs],
});

export type OccurCmdArgs = AddCmdArgs & TimeArgs;

export async function occurCmd(
  args: OccurCmdArgs | string | string[],
  preparsed?: Partial<OccurCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(occurCmdArgs, args, preparsed);
  const { time: occurTime } = handleTimeArgs(args);
  if (occurTime !== undefined) {
    args.cmdData ??= {};
    args.cmdData.occurTime = occurTime;
  }

  return await addCmd(args);
}
