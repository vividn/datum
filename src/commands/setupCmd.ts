import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
