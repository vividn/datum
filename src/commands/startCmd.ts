import { Argv } from "yargs";
import { EitherDocument } from "../documentControl/DatumDocument";
import { switchCmd } from "./switchCmd";
import { occurArgs, OccurCmdArgs } from "./occurCmd";
import { DurationArgs, durationArgs } from "../input/durationArgs";

export const command = "start <field> [duration] [data..]";
export const desc = "add a start document";

export function durOccurArgs(yargs: Argv): Argv {
  return durationArgs(occurArgs(yargs));
}
export function builder(yargs: Argv): Argv {
  return durOccurArgs(yargs);
}

export type StartCmdArgs = OccurCmdArgs & DurationArgs;

export async function startCmd(args: StartCmdArgs): Promise<EitherDocument> {
  return await switchCmd({ ...args, state: true });
}
