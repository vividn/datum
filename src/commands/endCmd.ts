import { Argv } from "yargs";
import { EitherDocument } from "../documentControl/DatumDocument";
import { durOccurArgs, StartCmdArgs } from "./startCmd";
import { switchCmd } from "./switchCmd";
import set from "lodash.set";

export const command = "end <field> [data..]";
export const desc = "add an end document";

export function builder(yargs: Argv): Argv {
  return durOccurArgs(yargs);
}

export type EndCmdArgs = StartCmdArgs;

export async function endCmd(args: StartCmdArgs): Promise<EitherDocument> {
  args.cmdData ??= {};
  set(args.cmdData, "state.id", false);
  return await switchCmd({ ...args });
}
