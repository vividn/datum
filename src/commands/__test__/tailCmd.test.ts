import {
  coloredChalk,
  delay,
  deterministicHumanIds,
  mockedLogLifecycle,
  mockSpecs,
  setNow,
  testDbLifecycle,
} from "../../__test__/test-utils";
import { endCmd } from "../endCmd";
import { startCmd } from "../startCmd";
import { switchCmd } from "../switchCmd";
import { occurCmd } from "../occurCmd";
import { tailCmd } from "../tailCmd";
import { Show } from "../../input/outputArgs";
import { setupCmd } from "../setupCmd";
import { addCmd } from "../addCmd";
import { updateCmd } from "../updateCmd";
import { DateTime, Settings } from "luxon";
import { generateSampleMorning } from "../../__test__/generateSampleMorning";
import { datumTimeToLuxon } from "../../time/datumTime";
import { DatumDocument } from "../../documentControl/DatumDocument";

const yesterday = "2023-10-15";
const today = "2023-10-16";
const tomorrow = "2023-10-17";

describe("tailCmd", () => {
  const { mockedLog } = mockedLogLifecycle();
  deterministicHumanIds();
  coloredChalk();
  const dbName = "tail_cmd_test";
  testDbLifecycle(dbName);
  mockSpecs();

  beforeEach(async () => {
    await setupCmd("");
    setNow(`20:00 ${today}`);
  });

  it("displays by default the last 10 occurrences in the database", async () => {
    await generateSampleMorning(today);
    const docs = await tailCmd("", { show: Show.Standard });
    expect(docs.length).toBe(10);
    expect(docs[0]._id).toMatchInlineSnapshot(`"run:2023-10-16T10:07:00.000Z"`);
    expect(docs.at(-1)?._id).toMatchInlineSnapshot(
      `"project:2023-10-16T11:45:00.000Z"`,
    );
    expect(mockedLog.mock.calls).toMatchSnapshot();
  });

  it("will display all if there are less than 10 occurrences in db", async () => {
    setNow(`8am ${today}`);
    await occurCmd("caffeine amount=100");
    setNow(`10am`);
    await switchCmd("project household");
    setNow(`10:30`);
    await endCmd("project");

    const docs = await tailCmd("", { show: Show.Standard });
    expect(docs.length).toBe(3);
    expect(mockedLog.mock.calls).toMatchSnapshot();
  });

  it("can display the last n occurrences", async () => {
    await generateSampleMorning(today);
    const docs = await tailCmd("-n 5");
    expect(docs.length).toBe(5);
  });

  it("will display all if there are less than n occurences in db", async () => {
    setNow(`8am ${today}`);
    await occurCmd("caffeine amount=100");
    setNow(`10am`);
    await switchCmd("project househould");
    setNow(`10:30`);
    await endCmd("project");

    const docs = await tailCmd("-n 5");
    expect(docs.length).toBe(3);
  });

  it("displays last occurrences of a specific field", async () => {
    await generateSampleMorning(today);
    const docs = (await tailCmd("stretch")) as DatumDocument[];
    expect(docs.length).toBe(4);
    await generateSampleMorning(yesterday);
    const docs2 = (await tailCmd("stretch")) as DatumDocument[];
    expect(docs2.length).toBe(8);
    expect(docs2.map((doc) => doc.data.field)).toEqual(
      Array(8).fill("stretch"),
    );
  });

  describe("tail metrics", () => {
    let occur1Id: string, occur2Id: string, occur3Id: string, addId: string;
    beforeEach(async () => {
      setNow("19:00");
      ({ _id: occur1Id } = await occurCmd("alcohol"));
      setNow("20:00");
      ({ _id: addId } = await addCmd(
        `person -b '{ name: "john doe", age: 35 }' --id %name`,
      ));
      setNow("21:00");
      ({ _id: occur3Id } = await endCmd("socialize"));
      setNow("+1");
      ({ _id: occur2Id } = await startCmd("socialize -t 1930"));

      // modify the doc without an occur time
      setNow("22:00");
      await updateCmd(`"${addId}" age=36`);
    });

    it("defaults to a hybrid display of occurTime and createTime", async () => {
      const docs = await tailCmd("", { show: Show.Standard });
      expect(docs.length).toBe(4);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        occur2Id,
        addId,
        occur3Id,
      ]);
      expect(mockedLog.mock.calls).toMatchSnapshot();

      const explicitHybridDocs = await tailCmd("--metric hybrid");
      expect(explicitHybridDocs).toEqual(docs);
    });

    it("can just display occurTime tail", async () => {
      const docs = await tailCmd("-m occur");
      expect(docs.length).toBe(3);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        occur2Id,
        occur3Id,
      ]);
      expect(mockedLog.mock.calls).toMatchSnapshot();
    });

    it("can display modifyTime tail", async () => {
      const docs = await tailCmd("-m modify");
      expect(docs.length).toBe(4);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        occur3Id,
        occur2Id,
        addId,
      ]);
      expect(mockedLog.mock.calls).toMatchSnapshot();
    });

    it("can display createTime tail", async () => {
      const docs = await tailCmd("-m create");
      expect(docs.length).toBe(4);
      expect(docs.map((doc) => doc._id)).toEqual([
        occur1Id,
        addId,
        occur3Id,
        occur2Id,
      ]);
    });
  });

  it("can display a tail from a certain moment in time", async () => {
    await generateSampleMorning(today);
    const docs1 = await tailCmd("-t 9:30");
    expect(docs1.length).toBe(8);
    expect(docs1.at(-1)?._id).toMatchInlineSnapshot(
      `"stretch:2023-10-16T09:30:00.000Z"`,
    );

    const docs2 = await tailCmd("-d yesterday -t 23:30");
    expect(docs2.length).toBe(1);

    const docs3 = await tailCmd("-yy -t 22");
    expect(docs3.length).toBe(0);
  });

  it("displays future entries if no specific time is given or --omit-timestamp is given", async () => {
    setNow(`20:00 ${today}`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);

    const docs = (await tailCmd("")) as DatumDocument[];
    const lastOccur = docs.at(-1)?.data.occurTime?.utc;
    expect(DateTime.fromISO(lastOccur ?? "").toISODate()).toEqual(tomorrow);

    const docsomitTimestamp = (await tailCmd("")) as DatumDocument[];
    const lastOccuromitTimestamp =
      docsomitTimestamp.at(-1)?.data.occurTime?.utc;
    expect(DateTime.fromISO(lastOccuromitTimestamp ?? "").toISODate()).toEqual(
      tomorrow,
    );
  });

  it("does not display future entries if now is given specifically as the time", async () => {
    setNow(`20:00 ${today}`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);

    const docs = (await tailCmd("-t now")) as DatumDocument[];
    const lastOccur = docs.at(-1)?.data.occurTime?.utc;
    expect(DateTime.fromISO(lastOccur ?? "").toISODate()).toEqual(today);
  });

  it("displays all occurrences on a day if date is given without time", async () => {
    await generateSampleMorning(today);
    await generateSampleMorning(yesterday);
    const docs = (await tailCmd("-d today")) as DatumDocument[];
    const yesterdayDocs = (await tailCmd("-y")) as DatumDocument[];

    expect(docs.length).toBeGreaterThan(10);
    expect(
      docs
        .map((doc) => doc.data.occurTime?.utc ?? "")
        .every(
          (occurTime) => DateTime.fromISO(occurTime).toISODate() === today,
        ),
    ).toBe(true);
    expect(
      yesterdayDocs
        .map((doc) => doc.data.occurTime?.utc ?? "")
        .every(
          (occurTime) => DateTime.fromISO(occurTime).toISODate() === yesterday,
        ),
    ).toEqual(true);
  });

  it("displays occurrences on a given day properly even when user is in a different timezone", async () => {
    // use New Zealand so that early events in the sample morning are previous day in UTC
    Settings.defaultZone = "Pacific/Auckland";
    setNow(`${today} 20:00`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);
    const docs = (await tailCmd(`-d ${today}`)) as DatumDocument[];
    expect(docs.length).toBeGreaterThan(10);

    // when mapping just from utc, not all should be on today
    const utcDates = docs.map((doc) => {
      const utc = doc.data.occurTime?.utc;
      if (!utc) {
        throw Error("bad occurTime");
      }
      return DateTime.fromISO(utc).toUTC().toISODate();
    });
    expect(utcDates.every((date) => date === today)).toBe(false);

    const localDates = docs.map((doc) =>
      datumTimeToLuxon(doc.data.occurTime)?.toISODate(),
    );
    expect(localDates.every((date) => date === today)).toBe(true);

    Settings.defaultZone = "system";
  });

  it("when requesting the entire date, have full day occurrences be at the top", async () => {
    Settings.defaultZone = "Pacific/Auckland";
    setNow(`${today} 20:00`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);
    const lengthWithoutFullDayDocs = (await tailCmd(`-d ${tomorrow}`)).length;

    const fullDayDoc1 = await occurCmd(`field -d ${tomorrow}`);
    const fullDayDoc2 = await occurCmd("otherField -d +1");
    const docs = await tailCmd(`-d ${tomorrow}`);
    expect(docs.length).toEqual(lengthWithoutFullDayDocs + 2);

    expect(docs.slice(0, 2)).toEqual([fullDayDoc1, fullDayDoc2]);
    Settings.defaultZone = "system";
  });

  it("displays only n latest occurrences on a day if date and -n is given", async () => {
    Settings.defaultZone = "Pacific/Auckland";
    setNow(`${today} 20:00`);
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);

    const fullDayDoc1 = await occurCmd(`field -d ${tomorrow}`);
    const fullDayDoc2 = await occurCmd(`otherField -d +1`);
    const allDocs = await tailCmd(`-d ${tomorrow}`);
    const limitedDocs = await tailCmd(`-d ${tomorrow} -n 5`);

    expect(limitedDocs.length).toEqual(5);
    expect(limitedDocs).not.toContainEqual(fullDayDoc1);
    expect(limitedDocs).not.toContainEqual(fullDayDoc2);
    expect(limitedDocs).toEqual(allDocs.slice(-5));
    Settings.defaultZone = "system";
  });

  it("can display a custom format for the tail commands", async () => {
    await generateSampleMorning(today);
    await generateSampleMorning(tomorrow);

    await occurCmd("", { show: Show.None });
    await tailCmd("--format-string ::%field%::");
    expect(mockedLog.mock.calls).toMatchInlineSnapshot(`
      [
        [
          "::run::
      ::stretch::
      ::environment::
      ::stretch::
      ::pushup::
      ::environment::
      ::caffeine::
      ::project::
      ::project::
      ::project::",
        ],
      ]
    `);
  });

  it("does not display anything when show is None", async () => {
    await generateSampleMorning(today);
    await tailCmd("", { show: Show.None });
    expect(mockedLog).not.toHaveBeenCalled();
  });

  it("can watch tail output", async () => {
    setNow(`12:00 ${today}`);
    await generateSampleMorning(today);
    tailCmd("-w", { show: Show.Standard });
    await delay(100);
    expect(mockedLog).toHaveBeenCalledTimes(1);
    expect(mockedLog.mock.calls[0]).toMatchSnapshot();
    mockedLog.mockClear();
    await delay(100);
    expect(mockedLog).not.toHaveBeenCalled();
    await occurCmd("pushup amount=10");
    setNow(`12:30`);
    await occurCmd("caffeine amount=100 -c coffee");
    await delay(100);
    expect(mockedLog).toHaveBeenCalledTimes(2);
    expect(mockedLog.mock.calls).toMatchSnapshot();

    // Note: the tail watch is automatically completed when the database is destoryed in the afterEach
  });

  it("can display extra data columns in the output", async () => {
    await generateSampleMorning(today);
    await tailCmd("--column amount,distance --column comment", {
      show: Show.Standard,
    });
    expect(mockedLog).toHaveBeenCalledTimes(1);
    expect(mockedLog.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "      time  field        state           dur  hid    amount  distance  comment 
      10:07:00[90m+0[39m  [48;2;165;49;8m[38;2;255;255;255mrun[39m[49m          [48;2;165;49;8m[38;2;255;255;255m [39m[49mend                 869r3          5.4               
      10:07:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m       [48;2;126;132;148m[38;2;0;0;0mstart[39m[49m[48;2;126;132;148m[38;2;0;0;0m [39m[49m              gs6pb                            
      10:15:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;195;1;99m[38;2;255;255;255m [39m[49m[48;2;16;106;108m[38;2;255;255;255mhome[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m               92g32                            
      10:15:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m      [48;2;126;132;148m[38;2;0;0;0m [39m[49mend                 chq8f                            
      10:18:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [48;2;136;136;136m[38;2;102;102;102m[38;2;148;83;231m●[39m[39m[49m                    getek  10                        
      10:30:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;16;106;108m[38;2;255;255;255m [39m[49m[48;2;195;1;99m[38;2;255;255;255moutside[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m       5m   2abgf                            
      11:00:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m     [48;2;136;136;136m[38;2;102;102;102m[38;2;231;36;248m●[39m[39m[49m                    cav6f  100               "coffee"
      11:00:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m       [48;2;175;103;202m[38;2;0;0;0memails[39m[49m[48;2;44;177;173m[38;2;0;0;0m,tasks[39m[49m[48;2;175;103;202m[38;2;44;177;173m▞[39m[49m       yafyz                            
      11:30:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m▞[39m[49m[48;2;212;135;136m[38;2;0;0;0mmeeting[39m[49m[48;2;175;103;202m[38;2;44;177;173m [39m[49m       10m  989x0                            
      11:45:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m▞[39m[49mend                 q9nko                            
      ",
      ]
    `);
  });
});
