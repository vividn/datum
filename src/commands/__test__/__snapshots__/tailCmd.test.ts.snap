// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tailCmd can watch tail output 1`] = `
[
  "      time  field        state           dur  hid  
10:07:00[90m+0[39m  [48;2;165;49;8m[38;2;255;255;255mrun[39m[49m          [48;2;165;49;8m[38;2;255;255;255m [39m[49mend                 869r3
10:07:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m       [48;2;126;132;148m[38;2;0;0;0mstart[39m[49m[48;2;126;132;148m[38;2;0;0;0m [39m[49m              gs6pb
10:15:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;195;1;99m[38;2;255;255;255m [39m[49m[48;2;16;106;108m[38;2;255;255;255mhome[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m               92g32
10:15:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m      [48;2;126;132;148m[38;2;0;0;0m [39m[49mend                 chq8f
10:18:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [48;2;136;136;136m[38;2;102;102;102m[38;2;148;83;231m‚óè[39m[39m[49m                    getek
10:30:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;16;106;108m[38;2;255;255;255m [39m[49m[48;2;195;1;99m[38;2;255;255;255moutside[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m       5m   2abgf
11:00:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m     [48;2;136;136;136m[38;2;102;102;102m[38;2;231;36;248m‚óè[39m[39m[49m                    cav6f
11:00:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m       [48;2;175;103;202m[38;2;0;0;0memails[39m[49m[48;2;44;177;173m[38;2;0;0;0m,tasks[39m[49m[48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m       yafyz
11:30:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m[48;2;212;135;136m[38;2;0;0;0mmeeting[39m[49m[48;2;175;103;202m[38;2;44;177;173m [39m[49m       10m  989x0
11:45:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49mend                 q9nko
",
]
`;

exports[`tailCmd can watch tail output 2`] = `
[
  [
    "      time  field        state           dur  hid  
10:07:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m       [48;2;126;132;148m[38;2;0;0;0mstart[39m[49m[48;2;126;132;148m[38;2;0;0;0m [39m[49m              gs6pb
10:15:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;195;1;99m[38;2;255;255;255m [39m[49m[48;2;16;106;108m[38;2;255;255;255mhome[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m               92g32
10:15:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m      [48;2;126;132;148m[38;2;0;0;0m [39m[49mend                 chq8f
10:18:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [48;2;136;136;136m[38;2;102;102;102m[38;2;148;83;231m‚óè[39m[39m[49m                    getek
10:30:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;16;106;108m[38;2;255;255;255m [39m[49m[48;2;195;1;99m[38;2;255;255;255moutside[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m       5m   2abgf
11:00:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m     [48;2;136;136;136m[38;2;102;102;102m[38;2;231;36;248m‚óè[39m[39m[49m                    cav6f
11:00:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m       [48;2;175;103;202m[38;2;0;0;0memails[39m[49m[48;2;44;177;173m[38;2;0;0;0m,tasks[39m[49m[48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m       yafyz
11:30:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m[48;2;212;135;136m[38;2;0;0;0mmeeting[39m[49m[48;2;175;103;202m[38;2;44;177;173m [39m[49m       10m  989x0
11:45:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49mend                 q9nko
12:00:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [38;2;148;83;231m[38;2;148;83;231m‚óè[39m[39m                    racga
",
  ],
  [
    "      time  field        state           dur  hid  
10:15:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;195;1;99m[38;2;255;255;255m [39m[49m[48;2;16;106;108m[38;2;255;255;255mhome[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m               92g32
10:15:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m      [48;2;126;132;148m[38;2;0;0;0m [39m[49mend                 chq8f
10:18:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [48;2;136;136;136m[38;2;102;102;102m[38;2;148;83;231m‚óè[39m[39m[49m                    getek
10:30:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;16;106;108m[38;2;255;255;255m [39m[49m[48;2;195;1;99m[38;2;255;255;255moutside[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m       5m   2abgf
11:00:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m     [48;2;136;136;136m[38;2;102;102;102m[38;2;231;36;248m‚óè[39m[39m[49m                    cav6f
11:00:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m       [48;2;175;103;202m[38;2;0;0;0memails[39m[49m[48;2;44;177;173m[38;2;0;0;0m,tasks[39m[49m[48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m       yafyz
11:30:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m[48;2;212;135;136m[38;2;0;0;0mmeeting[39m[49m[48;2;175;103;202m[38;2;44;177;173m [39m[49m       10m  989x0
11:45:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49mend                 q9nko
12:00:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [38;2;148;83;231m[38;2;148;83;231m‚óè[39m[39m                    racga
12:30:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m     [38;2;231;36;248m[38;2;231;36;248m‚óè[39m[39m                    wwtyu
",
  ],
]
`;

exports[`tailCmd displays by default the last 10 occurrences in the database 3`] = `
[
  [
    "      time  field        state           dur  hid  
10:07:00[90m+0[39m  [48;2;165;49;8m[38;2;255;255;255mrun[39m[49m          [48;2;165;49;8m[38;2;255;255;255m [39m[49mend                 869r3
10:07:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m       [48;2;126;132;148m[38;2;0;0;0mstart[39m[49m[48;2;126;132;148m[38;2;0;0;0m [39m[49m              gs6pb
10:15:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;195;1;99m[38;2;255;255;255m [39m[49m[48;2;16;106;108m[38;2;255;255;255mhome[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m               92g32
10:15:00[90m+0[39m  [48;2;126;132;148m[38;2;0;0;0mstretch[39m[49m      [48;2;126;132;148m[38;2;0;0;0m [39m[49mend                 chq8f
10:18:00[90m+0[39m  [48;2;148;83;231m[38;2;255;255;255mpushup[39m[49m       [48;2;136;136;136m[38;2;102;102;102m[38;2;148;83;231m‚óè[39m[39m[49m                    getek
10:30:00[90m+0[39m  [48;2;233;0;228m[38;2;255;255;255menvironment[39m[49m  [48;2;16;106;108m[38;2;255;255;255m [39m[49m[48;2;195;1;99m[38;2;255;255;255moutside[39m[49m[48;2;16;106;108m[38;2;255;255;255m [39m[49m       5m   2abgf
11:00:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m     [48;2;136;136;136m[38;2;102;102;102m[38;2;231;36;248m‚óè[39m[39m[49m                    cav6f
11:00:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m       [48;2;175;103;202m[38;2;0;0;0memails[39m[49m[48;2;44;177;173m[38;2;0;0;0m,tasks[39m[49m[48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m       yafyz
11:30:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49m[48;2;212;135;136m[38;2;0;0;0mmeeting[39m[49m[48;2;175;103;202m[38;2;44;177;173m [39m[49m       10m  989x0
11:45:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m      [48;2;175;103;202m[38;2;44;177;173m‚ñû[39m[49mend                 q9nko
",
  ],
]
`;

exports[`tailCmd tail metrics can display modifyTime tail 1`] = `
[
  [
    "       time  field      state      hid  
[90mm[39m19:00:00[90m+0[39m  [48;2;186;191;136m[38;2;0;0;0malcohol[39m[49m    [48;2;136;136;136m[38;2;102;102;102m[38;2;186;191;136m‚óè[39m[39m[49m          elfuv
[90mm[39m21:00:00[90m+0[39m  [48;2;111;238;105m[38;2;0;0;0msocialize[39m[49m  [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49mend       d7j7m
[90mm[39m21:01:00[90m+0[39m  [48;2;111;238;105m[38;2;0;0;0msocialize[39m[49m  [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49m[48;2;111;238;105m[38;2;0;0;0mstart[39m[49m[48;2;111;238;105m[38;2;0;0;0m [39m[49m    w1fla
[90mm[39m22:00:00[90m+0[39m  [48;2;139;10;68m[38;2;255;255;255mperson[39m[49m     [38;2;139;10;68m¬¢[39m          en61v
",
  ],
]
`;

exports[`tailCmd tail metrics can just display occurTime tail 1`] = `
[
  [
    "      time  field      state      hid  
19:00:00[90m+0[39m  [48;2;186;191;136m[38;2;0;0;0malcohol[39m[49m    [48;2;136;136;136m[38;2;102;102;102m[38;2;186;191;136m‚óè[39m[39m[49m          elfuv
19:30:00[90m+0[39m  [48;2;111;238;105m[38;2;0;0;0msocialize[39m[49m  [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49m[48;2;111;238;105m[38;2;0;0;0mstart[39m[49m[48;2;111;238;105m[38;2;0;0;0m [39m[49m    w1fla
21:00:00[90m+0[39m  [48;2;111;238;105m[38;2;0;0;0msocialize[39m[49m  [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49mend       d7j7m
",
  ],
]
`;

exports[`tailCmd tail metrics defaults to a hybrid display of occurTime and createTime 1`] = `
[
  [
    "       time  field      state      hid  
 19:00:00[90m+0[39m  [48;2;186;191;136m[38;2;0;0;0malcohol[39m[49m    [48;2;136;136;136m[38;2;102;102;102m[38;2;186;191;136m‚óè[39m[39m[49m          elfuv
 19:30:00[90m+0[39m  [48;2;111;238;105m[38;2;0;0;0msocialize[39m[49m  [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49m[48;2;111;238;105m[38;2;0;0;0mstart[39m[49m[48;2;111;238;105m[38;2;0;0;0m [39m[49m    w1fla
[90mc[39m20:00:00[90m+0[39m  [48;2;139;10;68m[38;2;255;255;255mperson[39m[49m     [38;2;139;10;68m¬¢[39m          en61v
 21:00:00[90m+0[39m  [48;2;111;238;105m[38;2;0;0;0msocialize[39m[49m  [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49mend       d7j7m
",
  ],
]
`;

exports[`tailCmd will display all if there are less than 10 occurrences in db 1`] = `
[
  [
    "      time  field     state          hid  
08:00:00[90m+0[39m  [48;2;231;36;248m[38;2;255;255;255mcaffeine[39m[49m  [48;2;136;136;136m[38;2;102;102;102m[38;2;231;36;248m‚óè[39m[39m[49m              elfuv
10:00:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m   [48;2;136;136;136m[38;2;102;102;102m‚àÖ[39m[49m[48;2;242;221;108m[38;2;0;0;0mhousehold[39m[49m[48;2;242;221;108m[38;2;0;0;0m [39m[49m    en61v
10:30:00[90m+0[39m  [48;2;70;248;111m[38;2;0;0;0mproject[39m[49m   [48;2;242;221;108m[38;2;0;0;0m [39m[49mend           d7j7m
",
  ],
]
`;
